buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt

```diff
@@ -118,6 +127,40 @@ class UnitTestAndCompilePlugin : Plugin<Project> {
         addCompileAllTask()
     }
 
+    private
+    fun Project.configureSourcesVariant() {
+        plugins.withType<JavaPlugin> {
+            extensions.getByType<JavaPluginExtension>().apply {
+                withSourcesJar()
+            }
+            val sourcesElements: Configuration by configurations.getting {
+                extendsFrom(configurations["implementation"])
+                outgoing.variants.create("sourcesDirectory") {
+                    attributes.attribute(Attribute.of("artifactType", String::class.java), "java-sources-directory")
+                    val sourceSet = the<SourceSetContainer>()[SourceSet.MAIN_SOURCE_SET_NAME]
+                    sourceSet.java.srcDirs.forEach {
+                        artifact(it)
+                    }
+                }
+            }
+
+            val sourcesPath: Configuration by configurations.creating {
+                isCanBeConsumed = false
+                extendsFrom(configurations["implementation"])
+                attributes {
+                    attribute(DocsType.DOCS_TYPE_ATTRIBUTE, objects.named(DocsType.SOURCES))
+                }
+            }
```

Comment:
Thinking about this again: This is rather specific to the ToolingAPI project. Other projects might not need to get the sources transitively. I think it would be better to move this part (the creation of `sourcesPath`) to `tooling-api.gradle.kts`.